### PR TITLE
Wallet manual tests fixes

### PR DESCRIPTION
**Pull request by bokobza**
**Should be reviewed commit by commit.**

These fixes bugs found during manually testing the wallet functionality.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 1]
```diff
@@ -408,6 +408,9 @@ public HdAccount GetUnusedAccount(Wallet wallet, string password)
 
                 // No unused account was found, create a new one.
                 account = wallet.AddNewAccount(password, this.coinType, this.dateTimeProvider.GetTimeOffset());
+                IEnumerable<HdAddress> newReceivingAddresses = account.CreateAddresses(this.network, this.walletSettings.UnusedAddressesBuffer);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Won't it be better to move this "creation" logic into `AddNewAccount`?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 2]
```diff
@@ -591,9 +594,11 @@ public AddressBalance GetAddressBalance(string address)
 
             lock (this.lockObject)
             {
+                HdAddress hdAddress = null;
+
                 foreach (Wallet wallet in this.Wallets)
                 {
-                    HdAddress hdAddress = wallet.GetAllAddressesByCoinType(this.coinType).FirstOrDefault(a => a.Address == address);
+                    hdAddress = wallet.GetAllAddressesByCoinType(this.coinType).FirstOrDefault(a => a.Address == address);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Maybe pass address into `GetAllAddressesByCoinType`, thereby creating an overload for the method?

**Comment by bokobza** [comment]
Added commit that fixes https://github.com/stratisproject/StratisBitcoinFullNode/issues/2417

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 2 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Not entirely sure that would make sense, given the name of the method and its presumed purpose.

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs [hunk 3]
```diff
@@ -328,7 +328,7 @@ public HdAccount GetFirstUnusedAccount()
             if (this.Accounts == null)
                 return null;
 
-            List<HdAccount> unusedAccounts = this.Accounts.Where(acc => !acc.ExternalAddresses.Any() && !acc.InternalAddresses.Any()).ToList();
+            List<HdAccount> unusedAccounts = this.Accounts.Where(acc => !acc.ExternalAddresses.SelectMany(add => add.Transactions).Any() && !acc.InternalAddresses.SelectMany(add => add.Transactions).Any()).ToList();
```
[the remark this answers is not in the document]
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: yes, an unused account is defined as an account that has no transaction against any of its addresses. Look here: https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery

We used to just look at the number of addresses, which was wrong.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 1 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: In theory yes. However, sometimes (like in cold staking), you want to just create an account and be in control of what you generate.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 2 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I think this overload would be confusing given the name of the function.

File: src/Stratis.Bitcoin.Features.Wallet/Wallet.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/Wallet.cs]: 👍

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: OK makes sense, I'm just worried that if someone calls `AddNewAccount` from anywhere and expect addresses to be in there, there won't be?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: `GetAllAddressesByCoinTypeAndAddress` ? LOL maybe that is too much. Anyway, all good then 👍

**Pull request merged**